sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/XPathCheck.java

```diff
@@ -152,4 +155,48 @@ public Iterator<String> getPrefixes(String val) {
       return null;
     }
   }
+
+  /**
+   * An internal re-implementation of the PrefixResolver from the xalan-j library.
+   */
+  private static class PrefixResolver {
+    Node mContext;
+
+    public PrefixResolver(Node xpathExpressionContext) {
+      mContext = xpathExpressionContext;
+    }
+
+    public String getNamespaceForPrefix(String prefix) {
+      return getNamespaceForPrefix(prefix, mContext);
+    }
+
+    public String getNamespaceForPrefix(String prefix, Node namespaceContext) {
+      if (prefix.equals("xml")) {
+        return "http://www.w3.org/XML/1998/namespace";
+      }
+
+      for (Node current = namespaceContext; current != null; current = current.getParentNode()) {
+        if (current.getNodeName().indexOf(prefix + ":") == 0) {
+          return current.getNamespaceURI();
+        }
+        NamedNodeMap attributes = current.getAttributes();
+        for (int i = 0; i < attributes.getLength(); i++) {
+          Node attr = attributes.item(i);
+          if (prefix.equals(extractPrefixFromAttribute(attr))) {
+            return attr.getNodeValue();
+          }
+        }
+      }
+      return null;
```

Comment:
maybe worth a comment that we can't reach there and why or @CheckForNull annotation